File: USACO/Silver/Python/USACO 2019 US Open Silver Fence Planning.py
import sys

sys.stdin = open('fenceplan.in', 'r')
sys.stdout = open('fenceplan.out', 'w')

input = sys.stdin.readline

class Cow:
    def __init__(self, x=0, y=0):
        self.x = x
        self.y = y

n, m = map(int, input().split())
cows = [Cow() for _ in range(n)]
for i in range(n):
    cows[i].x, cows[i].y = map(int, input().split())

adj = [[] for _ in range(n)]
for _ in range(m):
    a, b = map(int, input().split())
    adj[a-1].append(b-1)
    adj[b-1].append(a-1)

min_per = float('inf')

visited = [False]*n
for i in range(n):
    if visited[i]:
        continue
    
    maxx = maxy = 0
    minx = miny = float('inf')
    stack = [i]
    while stack:
        cur = stack.pop()
        visited[cur] = True
        maxx = max(maxx, cows[cur].x)
        minx = min(minx, cows[cur].x)
        maxy = max(maxy, cows[cur].y)
        miny = min(miny, cows[cur].y)
        for j in adj[cur]:
            if not visited[j]:
                stack.append(j)
    
    min_per = min(min_per, 2*(maxx - minx) + 2*(maxy - miny))
    
print(min_per)

File: USACO/Silver/C++/USACO 2019 US Open Silver Fence Planning.cpp
#include <bits/stdc++.h>
using namespace std;

#define ll long long
#define MOD(a, b) ((((a) % (b)) + (b)) % (b))

bool cbs(pair<int, int> a, pair<int, int> b) {
    return a.second < b.second;
}


signed main() {
    freopen("fenceplan.in", "r", stdin);
    freopen("fenceplan.out", "w", stdout);

    cin.tie(NULL) -> sync_with_stdio(false);
    
    int n, m;
    cin >> n >> m;
    vector<pair<int, int>> cows(n);
    for(int i=0; i<n; ++i) cin >> cows[i].first >> cows[i].second;
    vector<vector<int>> adj(n);
    for(int i=0; i<m; ++i){
        int a, b;
        cin >> a >> b;
        --a; --b;
        adj[a].push_back(b);
        adj[b].push_back(a);
    }

    ll ans = 4000000000;
    vector<bool> visited(n);
    for(int i=0; i<n; ++i){
        if (visited[i]) continue;
        stack<int> st;
        st.push(i);
        int minx = cows[i].first;
        int maxx = cows[i].first;
        int miny = cows[i].second;
        int maxy = cows[i].second;
        while(!st.empty()){
            int cur = st.top();
            st.pop();
            visited[cur] = true;
            minx = min(minx, cows[cur].first);
            maxx = max(maxx, cows[cur].first);
            miny = min(miny, cows[cur].second);
            maxy = max(maxy, cows[cur].second);
            for(auto x: adj[cur]) if (!visited[x]) st.push(x);
        }
        ans = min(ans, 2LL*(maxy-miny)+2LL*(maxx-minx));
    }
    cout << ans;
}
